### PR TITLE
CORS-3284: IBMCloud: CAPI PreProvisioining

### PR DESCRIPTION
**Pull request by cjschaef**
Add PreProvision logic for IBM Cloud CAPI support, which can:
1. Provision ResourceGroup
2. Create COS Instance and Bucket
3. Upload RHCOS image to COS
4. Create IAM Authorization for VPC to COS

Related: https://issues.redhat.com/browse/CORS-3284

### CODE REVIEW

**Comment by cjschaef** [comment]
Testing performed using this hook, the IBM Cloud CAPI controller is still a work in progress, not ready for use for VPC.
#### Resource Group
```
# ibmcloud resource group eu-de-capi-test-23-sn62j --output JSON
[
    {
        "id": "d3c7a47acc584a11a6e06a43cbb4518f",
        "crn": "crn:v1:bluemix:public:resource-controller::..........::resource-group:d3c7a47acc584a11a6e06a43cbb4518f",
        ...
        "created_at": "2024-04-05T16:04:05.533Z",
        "updated_at": "2024-04-05T16:04:05.533Z"
    }
]
```


#### COS Instance
```
# ibmcloud resource service-instance eu-de-capi-test-23-sn62j-cos --output json | jq -r '.[0] | .name,.crn'
eu-de-capi-test-23-sn62j-cos
crn:v1:bluemix:public:cloud-object-storage:global......:66d2fa43-d16a-4058-bf8f-6948a4b03fd5::
```


#### RHCOS file (COS Object)
```
# ibmcloud cos objects --bucket eu-de-capi-test-23-sn62j-vsi-imge --region eu-de
OK
Found 1 object in bucket 'eu-de-capi-test-23-sn62j-vsi-imge':

Name                                                Last Modified              Object Size   
rhcos-416.94.202403071059-0-ibmcloud.x86_64.qcow2   Apr 05, 2024 at 16:04:43   2.22 GiB 
```


### IAM Auth
```
# ibmcloud iam authorization-policies | grep -A1 -B5 crn:v1:bluemix:public:cloud-object-storage:global......:66d2fa43-d16a-4058-bf8f-6948a4b03fd5::
ID:                        3e59823c-3a31-414a-82ad-50661de42611
Source service name:       is
Source service instance:   All instances
Source resource type:      image
Target service name:       cloud-object-storage
Target service instance:   crn:v1:bluemix:public:cloud-object-storage:global......:66d2fa43-d16a-4058-bf8f-6948a4b03fd5::
Roles:                     Reader
```

**Comment by rvanderp3** [comment]
@cjschaef can you split the vendor in to its own commit?

**Comment by cjschaef** [comment]
Shoot, yes I will do that.

**Comment by cjschaef** [comment]
The COS/IAM configuration does not appear to allow for VPC Custom Image access.
I am investigating the issue further to see why this isn't working as expected. Will update PR if/when I resolve the issue.

/hold

**Comment by cjschaef** [comment]
Ah, typo was the actual issue
```
--- a/pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go
+++ b/pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go
@@ -85,7 +85,7 @@ func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionIn
                        return fmt.Errorf("failed creating RHCOS image COS instance: %w", err)
                }
        }
-       bucketName := fmt.Sprintf("%s-vsi-imge", in.InfraID)
+       bucketName := fmt.Sprintf("%s-vsi-image", in.InfraID)
```

With the fixed typo, the permissions work fine and a VPC Custom Image is successfully created (by using my branched IBM Cloud CAPI)
```
# ic is image eu-de-capi-test-39-d5npk-rhcos -q
                                  
ID                             r010-4100d484-a4c7-4887-b931-3c707aeec74f   
Name                           eu-de-capi-test-39-d5npk-rhcos   
CRN                            crn:v1:bluemix:public:is:eu-de:a/<foo>::image:r010-4100d484-a4c7-4887-b931-3c707aeec74f   
Status                         available   
Operating system               Name                       Architecture   Vendor    Version   Dedicated host only      
                               rhel-coreos-stable-amd64   amd64          Red Hat   stable    false      
                                  
Created                        2024-04-17T16:51:54+00:00   
Visibility                     private   
Owner type                     user   
Minimum provisioned size(GB)   100   
SHA256 Checksum                foo   
File size(GB)                  3   
Encryption                     none   
Resource group                 foo
Resource type                  image   
Catalog Offering Managed       false  
```

I'm going to make some additional changes to bring the COS resources more in line with those created by Terraform (Service Plan and Storage class).

**Comment by cjschaef** [comment]
I should have resolved the items above and this is ready for review once more.

/unhold

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 1]
```diff
@@ -20,5 +26,90 @@ func (p Provider) Name() string {
...
+		// We could create the Resource Group, defined by user, but that might make resource cleanup more difficult.
+		if in.InstallConfig.Config.Platform.IBMCloud.ResourceGroupName != "" {
+			return fmt.Errorf("provided resource group not found: %w", err)
+		}
+	}
+
+	// Create Resource Group if it wasn't found (and isn't expected to be an existing RG).
+	if resourceGroup == nil {
+		err := client.CreateResourceGroup(ctx, resourceGroupName)
+		if err != nil {
```
Comment (r4f4) [inline pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go:65]:
```suggestion
		if err := client.CreateResourceGroup(ctx, resourceGroupName); err != nil {
```

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 2]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
 	return nil
 }
 
+// CreateCOSBucket will create a new COS Bucket in the COS Instance, based on the Instance ID.
+func (c *Client) CreateCOSBucket(ctx context.Context, cosInstanceID string, bucketName string, region string) error {
+	_, cancel := context.WithTimeout(ctx, 1*time.Minute)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Sorry, copied too much from other functions. Will remove or resolve in these new functions.

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 3]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
...
+	locationConstraint := fmt.Sprintf("%s-%s", region, "smart")
+
+	options := &ibms3.CreateBucketInput{
+		Bucket: ptr.To(bucketName),
+		CreateBucketConfiguration: &ibms3.CreateBucketConfiguration{
+			LocationConstraint: ptr.To(locationConstraint),
+		},
+	}
+	_, err := cosClient.CreateBucket(options)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Consolidated

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 4]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
...
+	cosClient := c.getCOSClient(cosInstanceID, region)
+
+	options := &ibms3.PutObjectInput{
+		Body:   cosaws.ReadSeekCloser(bytes.NewReader(sourceData)),
+		Bucket: cosaws.String(bucketName),
+		Key:    cosaws.String(fileName),
+	}
+
+	_, err := cosClient.PutObject(options)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Consolidated

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 5]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
...
+		return fmt.Errorf("failed collecting account ID: %w", err)
+	}
+
+	policyRoles := []iampolicymanagementv1.Roles{}
+
+	for _, role := range roles {
+		newRole := iampolicymanagementv1.Roles{
+			RoleID: ptr.To(role),
+		}
+		policyRoles = append(policyRoles, newRole)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Consolidated

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 6]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
...
+	if err != nil {
+		return fmt.Errorf("failed creation IAM policy management service: %w", err)
+	}
+
+	options := iamPolicyManagementService.NewCreateV2PolicyOptions(policyControl, "authorization")
+	options.SetSubject(policySubject)
+	options.SetResource(policyResource)
+
+	_, _, err = iamPolicyManagementService.CreateV2Policy(options)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Consolidated

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 7]
```diff
@@ -145,6 +165,185 @@ func (c *Client) loadSDKServices() error {
...
+		return fmt.Errorf("failed retrieving account ID: %w", err)
+	}
+
+	createRGOptions := c.managementAPI.NewCreateResourceGroupOptions()
+	createRGOptions.SetName(rgName)
+	createRGOptions.SetAccountID(*apiKeyDetails.AccountID)
+
+	// Create the Resource Group
+	_, _, err = c.managementAPI.CreateResourceGroup(createRGOptions)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Consolidated.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 8]
```diff
@@ -20,5 +26,90 @@ func (p Provider) Name() string {
 
 // PreProvision creates the IBM Cloud objects required prior to running capibmcloud.
 func (p Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionInput) error {
+	_, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: That's probably a good idea, will remove.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go (hunk 1 above)
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: Consolidated.

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 9]
```diff
@@ -145,6 +165,176 @@ func (c *Client) loadSDKServices() error {
...
+	return nil
+}
+
+// CreateResourceGroup creates a new IBM Cloud Resource Group.
+func (c *Client) CreateResourceGroup(ctx context.Context, rgName string) error {
+	localContext, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	// Get the Account ID
+	apiKeyDetails, err := c.GetAuthenticatorAPIKeyDetails(ctx)
```
Comment (r4f4) [inline pkg/asset/installconfig/ibmcloud/client.go:321]:
```suggestion
	apiKeyDetails, err := c.GetAuthenticatorAPIKeyDetails(localContext)
```

**Review by r4f4 — COMMENTED** [review]
/approve

/hold
because I think there is a typo in the usage of one of the contexts.

File: pkg/asset/installconfig/ibmcloud/client.go (hunk 9 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: sorry, yep, have fixed that.

**Comment by cjschaef** [comment]
Thanks.
But, I also realized I missed some of the go vendor files after I rebased and updated my non-go mod commit. Fixed that I think now.

I can rebase once #8291 merges as well.

**Comment by cjschaef** [comment]
I will try to get these go mod fixes resolved soon.

**Comment by cjschaef** [comment]
`e2e-ibmcloud-ovn` MonitorTests failures appear to be related to https://issues.redhat.com/browse/OCPBUGS-30267

**Comment by r4f4** [comment]
`e2e-ibmcloud-ovn` is still terraform-based.

/test altinfra-e2e-ibmcloud-capi-ovn

**Comment by r4f4** [comment]
```
level=info msg=Consuming Install Config from target directory
level=fatal msg=failed to fetch Cluster API Manifests: failed to generate asset "Cluster API Manifests": unsupported platform "ibmcloud"
Create manifests exit code: 1
```
@cjschaef you're missing the CAPI cluster asset generation for `ibmcloud` [here](https://github.com/openshift/installer/blob/master/pkg/asset/manifests/clusterapi/cluster.go#L148} as well as the CAPI machines asset [here]( https://github.com/openshift/installer/blob/master/pkg/asset/machines/clusterapi.go#L444)

**Comment by cjschaef** [comment]
@r4f4 We do not have the support we need in CAPI as of yet for the Infrastructure resources (Cluster) and the Machines.

The current implementation in CAPI for VPC is not sufficient and needs further work. So, for now, I have only added the `PreProvision` code, to do work to setup prior to CAPI being run. The 4 items currently targeted for this PreProvision work are in the commit message, which this PR is designed to cover for now.

    1. Provision ResourceGroup
    2. Create COS Instance and Bucket
    3. Upload RHCOS image to COS
    4. Create IAM Authorization for VPC to COS

I hope to follow up with a PR(s) to add the Cluster manifest generation for Infrastructure `Provision` and later for Machine manifest generation, once we get some/most of that code into the IBM Cloud CAPI.

**Comment by r4f4** [comment]
> I hope to follow up with a PR(s) to add the Cluster manifest generation for Infrastructure `Provision` and later for Machine manifest generation, once we get some/most of that code into the IBM Cloud CAPI.

You'll never reach the PreProvision stage if you don't create the cluster/machine manifests (`create manifests` target is a dependency of `create cluster`). I understand you can't actually create the cluster/machine resources themselves, but you have to write out the manifests to at least be able to get to the PreProvision stage.

/approve cancel

**Comment by cjschaef** [comment]
I don't expect this code to be executed currently. Unless it needs to be in order to be merged. I had planned to follow up with additional PR's that would implement functionality that will not be executable (such as Ignition as well), until I can get CAPI implemented enough to generate the manifests.

Otherwise, I can add some generic manifest generation so that this code is executable, but still expect the CAPI controller will die horribly. And I expect the Cluster manifest content will need significant changes once the CAPI API's are fully updated.

Or, if you think it is better that these kind of changes wait until CAPI is implemented enough to reach this code (and thus test it), I can hold until then. It just might be significantly delayed.

**Comment by r4f4** [comment]
> I don't expect this code to be executed currently. Unless it needs to be in order to be merged. I had planned to follow up with additional PR's that would implement functionality that will not be executable (such as Ignition as well), until I can get CAPI implemented enough to generate the manifests.

You don't need CAPI to generate the manifests; you need it to consume those manifests and create the resources.

> Otherwise, I can add some generic manifest generation so that this code is executable, but still expect the CAPI controller will die horribly. And I expect the Cluster manifest content will need significant changes once the CAPI API's are fully updated.

That's fine and it's how many of the CAPI providers were implemented in the Installer. I thinks it's also easier for reviewers if they can see the code in action. If you can add just enough manifest definition to get the `PreProvision` to work, we will be a lot more confident about these PR changes and you'll know what to debug when new parts are introduced and things break.

Just make it clear in the PRs you open what is and what is not expected to work.

> Or, if you think it is better that these kind of changes wait until CAPI is implemented enough to reach this code (and thus test it), I can hold until then. It just might be significantly delayed.

From my experience with other providers, you end up uncovering missing parts as you implement the required features and you see issues crop up in CI and local testing. So waiting is definitely worse when we know there is work that has to happen in the Installer before we even get to the point where we rely on CAPI.

**Comment by r4f4** [comment]
> I don't expect this code to be executed currently. Unless it needs to be in order to be merged. 

It doesn't _need_ to but in this case it's so simple to add the missing bits that I don't see why we wouldn't do it.

**Comment by cjschaef** [comment]
@r4f4 Okay, I added the Cluster and Machine manifest generation in a generic form for now. Until we complete the work get get the IBM Cloud CAPI API's updated properly for VPC support for IPI. Then, we can pull in the new API's and refactor these manifests properly.

I don't expect the IBM Cloud VPC CAPI controller to function properly even with these manifests however.

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 10]
```diff
@@ -146,6 +166,176 @@ func (c *Client) loadSDKServices() error {
...
+	serviceInstanceKeyPtr := ptr.To("serviceInstance")
+	serviceNameKeyPtr := ptr.To("serviceName")
+	stringEqualsOperatorPtr := ptr.To("stringEquals")
+
+	apiKeyDetails, err := c.GetAuthenticatorAPIKeyDetails(ctx)
+	if err != nil {
+		return fmt.Errorf("failed collecting account ID: %w", err)
+	}
+
+	policyRoles := []iampolicymanagementv1.Roles{}
```
Comment (r4f4) [inline pkg/asset/installconfig/ibmcloud/client.go:233]: Nit: you can pre-allocate this array:
```suggestion
	policyRoles := make([]iampolicymanagementv1.Roles, 0, len(roles))
```

File: pkg/asset/manifests/ibmcloud/cluster.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+		},
+		Spec: capibmcloud.IBMVPCClusterSpec{
+			ControlPlaneEndpoint: capi.APIEndpoint{
+				Host: fmt.Sprintf("api.%s.%s", installConfig.Config.ObjectMeta.Name, installConfig.Config.BaseDomain),
+				Port: 6443,
+			},
+			Region:        platform.Region,
+			ResourceGroup: resourceGroup,
+		},
+	}
```
Comment (r4f4) [inline pkg/asset/manifests/ibmcloud/cluster.go:45]:
```suggestion
	}
    ibmcloudCluster.SetGroupVersionKind(capibmcloud.GroupVersion.WithKind("IBMVPCCluster"))
```

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 12]
```diff
@@ -24,5 +29,86 @@ func (Provider) BootstrapHasPublicIP() bool { return false }
...
+	_, err = client.GetCOSBucketByName(ctx, *cosInstance.ID, bucketName, region)
+	if err != nil {
+		err = client.CreateCOSBucket(ctx, *cosInstance.ID, bucketName, region)
+		if err != nil {
+			return fmt.Errorf("failed creating RHCOS image COS bucket: %w", err)
+		}
+	}
+
+	// Upload the RHCOS image to the COS Bucket.
+	cachedImage, err := cache.DownloadImageFile(string(*in.RhcosImage), cache.InstallerApplicationName)
```
Comment (r4f4) [inline pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]:
```suggestion
	cachedImage, err := cache.DownloadImageFile(in.RhcosImage.ControlPlane, cache.InstallerApplicationName)
```

Comment (r4f4) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: This asset has recently changed to hold both controlPlane and compute RHCOS images for the multi-arch cluster case. You also need to rebase this PR on top of current master.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go [hunk 13]
```diff
@@ -24,5 +29,86 @@ func (Provider) BootstrapHasPublicIP() bool { return false }
...
+	_, err = client.GetCOSBucketByName(ctx, *cosInstance.ID, bucketName, region)
+	if err != nil {
+		err = client.CreateCOSBucket(ctx, *cosInstance.ID, bucketName, region)
+		if err != nil {
+			return fmt.Errorf("failed creating RHCOS image COS bucket: %w", err)
+		}
+	}
+
+	// Upload the RHCOS image to the COS Bucket.
+	cachedImage, err := cache.DownloadImageFile(in.RhcosImage.Name(), cache.InstallerApplicationName)
```
Comment (r4f4) [inline pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go:94]:
```suggestion
	cachedImage, err := cache.DownloadImageFile(in.RhcosImage.ControlPlane, cache.InstallerApplicationName)
```
The `Name()` of the asset is just "Image" which is not what you want here.

File: pkg/asset/manifests/ibmcloud/cluster.go (hunk 11 above)
Comment (cjschaef) [reply pkg/asset/manifests/ibmcloud/cluster.go:45]: Added.

File: pkg/asset/installconfig/ibmcloud/client.go (hunk 10 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Updated.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go (hunk 12 above)
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: Updated and rebased.

File: pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go (hunk 13 above)
Comment (cjschaef) [reply pkg/infrastructure/ibmcloud/clusterapi/clusterapi.go]: updated per other comment.

**Review by r4f4 — COMMENTED** [review]
It seems to be working as in the second install attempt we get the error:
```
level=error msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed during pre-provisioning: failed creating RHCOS image COS bucket: failed to create cos bucket: BucketAlreadyExists: The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again.
```

It would be good to add some logs in the PreProvision  as resources are created, even if just `logrus.Debug` logs so we can tell which steps succeeded.

**Comment by r4f4** [comment]
Not sure what's causing 
```
level=error msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to create infrastructure manifest: no matches for kind "IBMVPCCluster" in version "infrastructure.cluster.x-k8s.io/v1beta2"
```
But solving this is out of scope for this PR.

**Comment by cjschaef** [comment]
I've added some log calls into PreProvision to help track progress.

I'm not sure what the delay is between the reattempts in the CI, but I'd guess it might be too quick that the COS Buckets are not completely deleted by the Infrastructure in time.

Perhaps there is a way to inject a longer delay in between attempts in the CI, as I expect this will be a common issue seen when attempt 1 fails for any reason.

---

I'm at a loss as to why the cluster manifest generation fails with
```
no matches for kind "IBMVPCCluster" in version "infrastructure.cluster.x-k8s.io/v1beta2"
```
but I'll worry about that elsewhere as you stated.

**Comment by r4f4** [comment]
> I'm at a loss as to why the cluster manifest generation fails with
> no matches for kind "IBMVPCCluster" in version "infrastructure.cluster.x-k8s.io/v1beta2"
> but I'll worry about that elsewhere as you stated.

Oooh, I think I know what is the issue: we need to run the ibmcloud CAPI provider as it's done for PowerVS: https://github.com/openshift/installer/blob/master/pkg/clusterapi/system.go#L286-L311

**Comment by r4f4** [comment]
> I'm not sure what the delay is between the reattempts in the CI, but I'd guess it might be too quick that the COS Buckets are not completely deleted by the Infrastructure in time.
> 
> Perhaps there is a way to inject a longer delay in between attempts in the CI, as I expect this will be a common issue seen when attempt 1 fails for any reason.

You can fix that around here https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh#L735-L773

**Comment by r4f4** [comment]
> > I'm at a loss as to why the cluster manifest generation fails with
> > no matches for kind "IBMVPCCluster" in version "infrastructure.cluster.x-k8s.io/v1beta2"
> > but I'll worry about that elsewhere as you stated.
> 
> Oooh, I think I know what is the issue: we need to run the ibmcloud CAPI provider as it's done for PowerVS: https://github.com/openshift/installer/blob/master/pkg/clusterapi/system.go#L286-L311

Here: https://github.com/openshift/installer/blob/master/pkg/clusterapi/system.go#L238-L239 :)

**Comment by cjschaef** [comment]
Should I add that change to this PR as well then?
https://github.com/openshift/installer/blob/master/pkg/clusterapi/system.go#L238-L239

Dang, I was hoping to keep this PR small, but guess it'll bloat getting generic values and setup in that will have to be replaced/fixed once CAPI implementation progresses.

**Comment by r4f4** [comment]
> > Here: https://github.com/openshift/installer/blob/master/pkg/clusterapi/system.go#L238-L239 :)
> 
> Should I add that change to this PR as well then?

I'll live it up to you. As far as this PR is concerned, we've validated it's pre-provisioning the resources.

**Comment by r4f4** [comment]
> Dang, I was hoping to keep this PR small, but guess it'll bloat getting generic values and setup in that will have to be replaced/fixed once CAPI implementation progresses.

Welcome to the "adding support for CAPI" club :)

**Comment by cjschaef** [comment]
Okay, I'll leave that implementation for a followup PR if that's okay then.

Yeah, the CAPI wrinkle has a lot of dependencies that in our case are still MIA, making things even more complex.
Thanks!

/retest

**Comment by cjschaef** [comment]
I think I've managed to track down more details on the `e2e-ibmcloud-ovn` failure
```
 [sig-auth][Feature:SCC][Early] should not have pod creation failures during install [Suite:openshift/conformance/parallel]
```

Think it will require some potential changes for the ibm-vpc-csi-driver-operator and how it applies RBAC resources versus the daemonset that fails temporarily.
https://issues.redhat.com/browse/OCPBUGS-42231

**Comment by cjschaef** [comment]
`e2e-ibmcloud-ovn` failure is due to
```

: [Jira:"Networking / router"] monitor test service-type-load-balancer-availability setup expand_less
{  failed during setup error waiting for replicaset: failed to create RC "service-test": replicationcontrollers "service-test" already exists}
```

This appears to be an issue with MonitorTests itself. I will see if a Jira is required to get this failure resolved (test setup failure).
Otherwise, I think the test are good to go.

**Comment by patrickdillon** [comment]
> `e2e-ibmcloud-ovn` failure is due to
> 
> ```
> 
> : [Jira:"Networking / router"] monitor test service-type-load-balancer-availability setup expand_less
> {  failed during setup error waiting for replicaset: failed to create RC "service-test": replicationcontrollers "service-test" already exists}
> ```
> 
> This appears to be an issue with MonitorTests itself. I will see if a Jira is required to get this failure resolved (test setup failure). This setup appears to hit issues in testing frequently: [openshift/cluster-cloud-controller-manager-operator#366 (comment)](https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/366#issuecomment-2377152526)
> 
> Otherwise, I think the test are good to go.

I think these errors are coming from tests using terraform installs. The `FeatureGateClusterAPIInstall` or `FeatureGateClusterAPIInstallIBMCloud` feature gate is needed to exercise this functionality. `FeatureGateClusterAPIInstall` is enabled in:

/test altinfra-e2e-ibmcloud-capi-ovn

**Comment by patrickdillon** [comment]
This LGTM. I have little experience with ibmcloud but reviewed as carefully as possible. 

Let's see what the new ci run turns up.

**Comment by cjschaef** [comment]
Looks like the new CI is failing, as the CAPI support for IBM Cloud is incomplete.
https://github.com/openshift/installer/pull/8235#issuecomment-2245621294
```
level=error msg=failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed to create infrastructure manifest: no matches for kind "IBMVPCCluster" in version "infrastructure.cluster.x-k8s.io/v1beta2"
```

There is still a lot of work to do to get IBM Cloud support for CAPI integrated into the installer.
I am hoping to follow up this PR with many more to complete that. Work on CAPI Infrastructure reconciliation is still a work in progress as well, as those PR's are being reviewed.

**Comment by patrickdillon** [comment]
> I am hoping to follow up this PR with many more to complete that. Work on CAPI Infrastructure reconciliation is still a work in progress as well, as those PR's are being reviewed.

Totally expected and fine. Just want to make sure the incremental changes work as expected.

**Comment by cjschaef** [comment]
I have opened a Jira for the `e2e-ibmcloud-ovn` test failure
https://issues.redhat.com/browse/OCPBUGS-42869

**Pull request merged**